src/ast.rs

```diff
@@ -44,240 +51,239 @@ lazy_static! {
-      if let Some(captures) = PATH_REFERENCE_RE.captures(&comment.text) {
-        let m = captures.get(1).unwrap();
-        references.push(TypeScriptReference::Path(
-          m.as_str().to_string(),
-          comment_match_to_source_range(comment, &m),
-        ));
-      } else if let Some(captures) = TYPES_REFERENCE_RE.captures(&comment.text)
-      {
-        let m = captures.get(1).unwrap();
-        references.push(TypeScriptReference::Types(
-          m.as_str().to_string(),
-          comment_match_to_source_range(comment, &m),
-        ));
+) -> Vec<SpecifierWithRange> {
+  // Analyze any JSDoc type imports
+  // We only analyze these on JavaScript types of modules, since they are
+  // ignored by TypeScript when type checking anyway and really shouldn't be
+  // there, but some people do strange things.
+  if !matches!(
+    parsed_source.media_type(),
+    MediaType::JavaScript | MediaType::Jsx | MediaType::Mjs | MediaType::Cjs
+  ) {
+    return Vec::new();
+  }
+
+  let mut deps = Vec::new();
+  for comment in parsed_source.comments().get_vec().iter() {
+    if comment.kind != CommentKind::Block || !comment.text.starts_with('*') {
+      continue;
+    }
+    for captures in JSDOC_IMPORT_RE.captures_iter(&comment.text) {
+      if let Some(m) = captures.get(1) {
+        deps.push(SpecifierWithRange {
+          text: m.as_str().to_string(),
+          range: comment_source_to_position_range(
+            comment.range().start,
+            &m,
+            parsed_source.text_info(),
+          ),
+        });
       }
     }
   }
-  references
+  deps
+}
+
+fn comment_source_to_position_range(
+  comment_start: SourcePos,
+  m: &Match,
+  text_info: &SourceTextInfo,
+) -> PositionRange {
+  // the comment text starts after the double slash or slash star, so add 2
+  let comment_start = comment_start + 2;
+  PositionRange {
+    start: Position::from_source_pos(comment_start + m.start(), text_info),
+    end: Position::from_source_pos(comment_start + m.end(), text_info),
+  }
 }
 
-/// Parses text to a `ParsedSource`.
-pub trait SourceParser {
-  /// Parses the provided module to a `ParsedSource`.
-  fn parse_module(
+/// A `ModuleAnalyzer` that caches the `deno_ast::ParsedSource` objects it creates.
+pub trait CapturingParsedSourceAnalyzer: ModuleAnalyzer {
+  fn parsed_source(
     &self,
     specifier: &ModuleSpecifier,
-    source: Arc<str>,
-    media_type: MediaType,
-  ) -> Result<ParsedSource, Diagnostic>;
+  ) -> Result<ParsedSource, anyhow::Error>;
 }
 
-// TODO(@dsherret) remove CapturingSourceParser
-
-/// An implementation of `SourceParser` that stores the parsed ASTs.
+/// Implementation of `CapturingParsedSourceAnalyzer` that stores the sources in a RefCell.
 #[derive(Default)]
-pub struct CapturingSourceParser {
-  modules: RefCell<HashMap<ModuleSpecifier, ParsedSource>>,
+pub struct RefCellCapturingParsedSourceAnalyzer {
+  sources: RefCell<HashMap<ModuleSpecifier, ParsedSource>>,
 }
```

Comment:
I found it was useful to implement these in deno_graph for resuse in downstream crates. Any ideas for better names?

Reply:
`DefaultAnalyzer`?

Reply:
That's much better. Thanks for the feedback and I think I have it much better now and it's been informed a bit more by me implementing this in the CLI now. Sorry for all the churn.